src/cli/setup_mode.rs

```diff
@@ -1,5 +1,6 @@
 use anyhow::Result;
 use clap::{builder::PossibleValuesParser, Parser};
+use itertools::Itertools;
```

Comment:
IMO this is a few different changes that warrant being in a separate commit:

1. Use itertools for `collect_vec()`
2. Import `ErrorKind`
3. Re-order the warning logging
4. Avoid unnecessary mapping

I'm also not a big fan of (1). IMO the use of `Itertools` is a little magic, and I don't feel like `collect_vec()` is all that much easier to understand than `collect::<Vec<_>>()`.

Given that this PR might take a while to be merged if we want to change everything to derive style, maybe these smaller changes should be in a separate PR that we can merge directly? (The same might go for the change to `maybe_official_toolchain_parser()`.)

Reply:
@djc Thanks for your timely reply! I'll split this into separate PRs soon.

Reply:
@djc I've made #3644. My plan for this commit in particular is that I'll
- Abandon `.collect_vec()` since you don't like it.
- Leave the `if` branch reorder here since it's meaningless other there.
- Put the rest over to #3644.

Reply:
PR rebased in accordance with the plan above.